src/caching_key_wrapper.rs

```diff
@@ -216,17 +214,14 @@ where
         Ok(key)
     }
 
-    async fn decrypt_data_key(
-        &self,
-        encrypted_key: &Vec<u8>,
-    ) -> Result<Key<U16>, KeyDecryptionError> {
+    async fn decrypt_data_key(&self, encrypted_key: &[u8]) -> Result<Key<S>, KeyDecryptionError> {
         if let Some(cached_key) = self.get_cached_decryption_key(encrypted_key).await {
             return Ok(cached_key);
         }
 
         let plaintext_key = self.provider.decrypt_data_key(encrypted_key).await?;
 
-        self.cache_decryption_key(encrypted_key, plaintext_key)
+        self.cache_decryption_key(encrypted_key, plaintext_key.clone())
```

Comment:
I'm looking at the current code in `main`, and it seems that the variable `plaintext_key` above, is actually `Key<U16>` which implements `Copy + Clone`. Although `.clone()` was not used previously, because it was `Copy`, I think a copy of this was in fact sent to the `self.cache_decryption_key(encrypted_key, plaintext_key)` as above.

In the current PR, `Key<S>` does not necessarily implement `Copy` hence the need for `.clone()`. I can add that constraint and this `.clone()` would go away. But of course, underlying copy still happens.

In any case, I feel like avoiding both cloning and copying the value here is not worth it for the following constraints:
- current methods in trait KeyProvider are expected to return owned values, and not just references
- current `encryption_cache` and `decryption_cache` have to own the values in the cache.

To me, these imply that the values need to exist in multiple copies anyway. We might able to try avoiding clones and copy, maybe the value stored in the cached needs to be wrapped individually in another smart pointer, and maybe `KeyProvider` trait methods should return smart pointer instead? I don't know how this works.

Reply:
Thank you for your thorough explanation @ragibkl.

FWIW, I'm ok with the clone. I suspect it has a very small impact in the grand scheme of things and we can micro-optimise when it makes sense to do so. I don't think we should be overly concerned about it now.

Reply:
Quick update. After quick pairing session with @coderdan , I made this so that `Key<S>: Copy`. That way we don't clone the values over anymore. It does get copied over, but it stays in the stack.